src/beliefpropagation/sqrt_beliefpropagation.jl

```diff
@@ -74,48 +67,49 @@ function sqrt_belief_propagation(
   return sqr_message_tensors(sqrt_mts)
 end
 
-function update_sqrt_message_tensor(
-  tn::ITensorNetwork, subgraph_vertices::Vector, sqrt_mts::Vector{ITensorNetwork};
-)
-  sqrt_mts_itensors = reduce(vcat, ITensor.(sqrt_mts); init=ITensor[])
-  v = only(unique(first.(subgraph_vertices)))
-  site_itensor = tn[v]
-  contract_list = ITensor[sqrt_mts_itensors; site_itensor]
+function sqrt_belief_propagation(ptn::PartitionedGraph; kwargs...)
+  mts = message_tensors(ptn)
+  return sqrt_belief_propagation(ptn, mts; kwargs...)
+end
+
+function update_sqrt_message_tensor(pψψ::PartitionedGraph, edge::PartitionEdge, sqrt_mts;)
+  v = only(filter(v -> v[2] == 1, vertices(pψψ, src(edge))))
+  site_itensor = unpartitioned_graph(pψψ)[v]
+  p_edges = setdiff(partitionedges(pψψ, boundary_edges(pψψ, [v]; dir=:in)), [reverse(edge)])
+  incoming_messages = [sqrt_mts[e_in] for e_in in p_edges]
+  incoming_messages = reduce(vcat, incoming_messages; init=ITensor[])
+  contract_list = ITensor[incoming_messages; site_itensor]
   contract_output = contract(
     contract_list; sequence=contraction_sequence(contract_list; alg="optimal")
   )
-  left_inds = [uniqueinds(contract_output, site_itensor); siteinds(tn, v)]
+  left_inds = [
+    uniqueinds(contract_output, site_itensor)
+    siteinds(unpartitioned_graph(pψψ), v)
+  ]
   Q, R = qr(contract_output, left_inds)
   normalize!(R)
-  return ITensorNetwork(R)
-end
-
-function update_sqrt_message_tensor(
-  tn::ITensorNetwork, subgraph::ITensorNetwork, mts::Vector{ITensorNetwork}; kwargs...
-)
-  return update_sqrt_message_tensor(tn, vertices(subgraph), mts; kwargs...)
+  return R
 end
 
 function sqrt_message_tensors(
-  ψ::ITensorNetwork,
-  mts::DataGraph;
-  eigen_message_tensor_cutoff=10 * eps(real(scalartype(ψ))),
-  regularization=10 * eps(real(scalartype(ψ))),
+  pψψ::PartitionedGraph,
+  mts;
+  eigen_message_tensor_cutoff=10 * eps(real(scalartype(unpartitioned_graph(pψψ)))),
+  regularization=10 * eps(real(scalartype(unpartitioned_graph(pψψ)))),
```

Comment:
I guess this kind of thing is unfortunate.

Ideally `scalartype` would be able to reach into any iterable object, and then it could "just work" for partitioned tensor networks. I started an issue about that here:  #127.